modules/iothubdeviceping/src/iothubdeviceping.c

```diff
@@ -0,0 +1,738 @@
+        amqpvalue_destroy(filter_set);
+    }
+    link_destroy(link);
+    session_destroy(session);
+    connection_destroy(connection);
+    xio_destroy(sasl_io);
+    xio_destroy(tls_io);
+    saslmechanism_destroy(sasl_mechanism_handle);
+    platform_deinit();
+
+    return result;
+}
+
+static int poll_eventhub_entry(IOTHUBDEVICEPING_HANDLE_DATA *handleData)
+{
+    const char *eventHubHost = ((const IOTHUBDEVICEPING_CONFIG *)(handleData->config))->EH_HOST;
+    const char *eventHubKeyName = ((const IOTHUBDEVICEPING_CONFIG *)(handleData->config))->EH_KEY_NAME;
+    const char *eventHubKey = ((const IOTHUBDEVICEPING_CONFIG *)(handleData->config))->EH_KEY;
+    const char *eventHubCompatibleName = ((const IOTHUBDEVICEPING_CONFIG *)(handleData->config))->EH_COMP_NAME;
+    const char *eventHubPartitionNum = ((const IOTHUBDEVICEPING_CONFIG *)(handleData->config))->EH_PARTITION_NUM;
+
+    int partitionNum = (int)atoi(((const IOTHUBDEVICEPING_CONFIG *)(handleData->config))->EH_PARTITION_NUM);
+
+    if (platform_init() != 0)
+    {
+        LogError("platform_init failed");
+        return -1;
+    }
+    else
+    {
+        MESSAGE_RECEIVER_CONTEXT *message_receiver_context_arr = (MESSAGE_RECEIVER_CONTEXT *)malloc(sizeof(MESSAGE_RECEIVER_CONTEXT) * partitionNum);
+        if (message_receiver_context_arr == NULL)
+        {
+            LogError("malloc returned NULL\r\n");
+            return -1;
+        }
+        else
+        {
+            char *result1 = concatenate("amqps://", eventHubHost);
+            char *result2 = concatenate("/", eventHubCompatibleName);
+            char *result3 = concatenate((const char *)result1, (const char *)result2);
+            char *result4 = concatenate((const char *)result3, "/ConsumerGroups/$Default/Partitions/");
+
+            // launch x num of threads for x num of partitions
+            int i = 0;
+            for (i = 0; i < partitionNum; i++)
+            {
+                int length = snprintf(NULL, 0, "%d", i);
+                char *str = malloc(length + 1);
+                snprintf(str, length + 1, "%d", i);
+                message_receiver_context_arr[i].address = (char *)concatenate((const char *)result4, (const char *)str);
+                message_receiver_context_arr[i].moduleHandleData = handleData;
+                message_receiver_context_arr[i].eventHubHost = eventHubHost;
+                message_receiver_context_arr[i].eventHubKeyName = eventHubKeyName;
+                message_receiver_context_arr[i].eventHubKey = eventHubKey;
+                message_receiver_context_arr[i].begin_execution_time = time(NULL);
+                message_receiver_context_arr[i].num_message_received = 0;
+                message_receiver_context_arr[i].message_received = false;
+                free(str);
+
+                if (ThreadAPI_Create(&message_receiver_context_arr[i].threadHandle, poll_eventhub_thread, &message_receiver_context_arr[i]) != THREADAPI_OK)
```

Comment:
Messages sent on behalf of a particular device *always* land in the same event hub partition. So you only need to listen on one partition.

Check out the [ResolvePartitionIndex](https://github.com/Azure/azure-iot-sdks/blob/master/c/testtools/iothub_test/src/iothubtest.c#L225) function, used by the device SDK's end-to-end tests. If you pass to it the device ID (string) and the number of partitions in the hub, it returns the partition index.